File: year2025/day3/part2/solution.py
#!/usr/bin/env python3
# Day 3: Lobby, Part 2

FILE_PATH = "../.input"

def main():
    total_joltage = 0
    num_batteries = 12

    with open(FILE_PATH, 'r') as banks_file:
        for bank in banks_file:
            bank = bank.strip()  # Clean list
            remaining_bank = list(map(int, bank))  # Convert to ints
            # print(f"DEBUG: bank: {bank}")
            batteries = remaining_bank[-num_batteries:]  # Fill with specified amount of batteries from end of list
            # print(f"DEBUG: starting batteries: {''.join(map(str,batteries))}")
            
            for cur_battery in remaining_bank[:-num_batteries][::-1]:
                # print(f"DEBUG: cur_battery: {cur_battery}")
                for i, battery in enumerate(batteries):
                    # print(f"DEBUG: {cur_battery}, {battery}")
                    if cur_battery < battery:
                        break
                    # print(f"DEBUG: Replacing {battery} -> {cur_battery}")
                    cur_battery, battery = battery, cur_battery
                    batteries[i] = battery
            
            combined_joltage = 0
            for joltage in batteries:
                combined_joltage = combined_joltage * 10 + joltage
            # print(f"DEBUG: {bank}: {combined_joltage}")
            total_joltage += combined_joltage
    print(f"Total Joltage: {total_joltage}")

if __name__ == "__main__":
    main()

File: year2025/day3/part2/solution.cpp
// Day 3: Lobby, Part 2

#include <iostream>
#include <fstream>
#include <string>
#include <algorithm>
#include <vector>

using namespace std;

const string FILE_PATH = "../.input";

int main() {
    unsigned long long total_joltage = 0;
    unsigned num_batteries = 12;

    ifstream banks_file(FILE_PATH);
    string bank;

    while (getline(banks_file, bank)) {
        // These are chars for display purposes, but behave the same with the exception of the ascii conversion when computing combined_joltage.
        vector<int> batteries;

        for (size_t i = bank.size() - num_batteries; i < bank.size(); ++i) {
            batteries.emplace_back(bank[i] - '0');  // Convert to int
        }

        //cout << "DEBUG: starting batteries: ";
        //for (int c : batteries) {
        //    cout << c;
        //}
        //cout << endl;

        string remaining_bank = bank.substr(0, bank.size() - num_batteries);  // Shrink bank
        reverse(remaining_bank.begin(), remaining_bank.end());  // Reverse bank

        for (char cur_battery_char : remaining_bank) {
            int cur_battery = cur_battery_char - '0';
            //cout << "DEBUG: cur_battery: " << cur_battery << endl;
            for (size_t j = 0; j < batteries.size(); ++j) {
                if (cur_battery < batteries[j]) {
                    break;
                }
                //cout << "DEBUG: Replacing " << batteries[j] << " -> " << cur_battery << endl;
                int temp_battery = batteries[j];
                batteries[j] = cur_battery;
                cur_battery = temp_battery;
            }
        }

        unsigned long long combined_joltage = 0;
        for (int joltage : batteries) {
            combined_joltage = (combined_joltage * 10) + joltage;
        }
        //cout << "DEBUG: " << bank << ": " << combined_joltage << endl;
        total_joltage += combined_joltage;
    }

    cout << "Total Joltage: " << total_joltage << endl;
    return 0;
}
